### PR TITLE
Smart contracts: Contract invocation

### PR DESCRIPTION
**Pull request by Zett98**
## Depends

- [x] #506
- [x] #515
## Problem
We need to have a way to invoke a contract
## Solution
add a user operation to invoke a contract and expose cli interface to it using already existing `transfer` command

### CODE REVIEW

**Review by anchpop — APPROVED** [review]

File: src/core/state.ml [hunk 1]
```diff
@@ -64,10 +64,32 @@ let apply_user_operation t user_operation =
...
+      Amount.
+        ( `Invocation_error "Not enought funds",
+          let comparison_result = compare balance invocation_cost in
+          comparison_result >= 0 ) in
 
+    let burn_cap = invocation_cost in
+    let initial_gas = Amount.(to_int (balance - burn_cap)) in
+    let wrap_error t = Result.map_error (fun x -> `Invocation_error x) t in
+    let%ok contract =
+      Contract_storage.get_contract t.contract_storage ~address:to_invoke
```
Comment (EduardoRFS) [inline src/core/state.ml:81]:
```suggestion
      Contract_storage.get_contract contract_storage ~address:to_invoke
```

File: src/core/state.ml [hunk 2]
```diff
@@ -64,10 +64,32 @@ let apply_user_operation t user_operation =
...
+      |> Option.fold ~none:(Error "Contract not found") ~some:Result.ok
+      |> wrap_error in
+    let%ok contract, _user_op_list =
+      Contract_vm.Interpreter.invoke ~arg:argument ~gas:initial_gas contract
+      |> wrap_error in
+    let contract_storage =
+      Contract_storage.update_contract_storage t.contract_storage
```
Comment (EduardoRFS) [inline src/core/state.ml:88]:
```suggestion
      Contract_storage.update_contract_storage contract_storage
```

File: src/core/state.ml [hunk 3]
```diff
@@ -64,10 +64,32 @@ let apply_user_operation t user_operation =
...
+        ~address:to_invoke ~updated_contract:contract in
+    Ok ({ contract_storage; ledger = t.ledger }, None)
```
Comment (EduardoRFS) [inline src/core/state.ml:90]:
```suggestion
    Ok ({ contract_storage; ledger }, None)
```

File: src/core/state.ml [hunk 4]
```diff
@@ -64,10 +64,32 @@ let apply_user_operation t user_operation =
...
 let apply_user_operation t user_operation =
   match apply_user_operation t user_operation with
   | Ok (t, receipt) -> (t, receipt)
-  (* TODO: use this origination_error for something *)
-  | Error (`Origination_error _) -> (t, None)
+  | Error (`Origination_error _ | `Invocation_error _) -> (t, None)
```
Comment (EduardoRFS) [inline src/core/state.ml:94]:
```suggestion
  (* TODO: use this erros for something *)
  | Error (`Origination_error _ | `Invocation_error _) -> (t, None)
```

File: src/core/contract_vm.mli [hunk 5]
```diff
@@ -10,3 +10,17 @@ end
...
+  type t [@@deriving yojson]
+  val make : arg:Yojson.Safe.t -> (t, string) result
+end
+
+module Interpreter : sig
+  val invoke :
+    Contract.t ->
+    arg:Invocation_payload.t ->
+    gas:int ->
+    (* @todo: unit should be user operation list *)
```
Comment (EduardoRFS) [inline src/core/contract_vm.mli:24]:
```suggestion
    (* @TODO: unit should be user operation list *)
```

File: src/core/contract_vm.ml [hunk 6]
```diff
@@ -14,6 +14,18 @@ module type VM = sig
...
+  module Invocation_payload : sig
+    type t [@@deriving yojson]
+    val make : arg:Yojson.Safe.t -> (t, string) result
+  end
+  module Interpreter : sig
+    val invoke :
+      Contract.t ->
+      arg:Invocation_payload.t ->
+      gas:int ->
+      (* @todo: unit should be user operation list *)
```
Comment (EduardoRFS) [inline src/core/contract_vm.ml:26]:
```suggestion
      (* @TODO: unit should be user operation list *)
```

File: src/core/contract_vm.ml [hunk 7]
```diff
@@ -89,6 +101,39 @@ module Lambda : VM = struct
...
+        "Invalid argument passed to the contract/Error within contract code"
+      | Interpreter_error (Undefined_variable | Over_applied_primitives) ->
+        "WOOOOOO!1!1! Bug within Lambda_vm interpreter"
+
+    let invoke contract ~(arg : Invocation_payload.t) ~gas =
+      let gas = Lambda_vm.Gas.make ~initial_gas:gas in
+      let%ok argument =
+        Raw_repr.Value.to_value ~gas arg.arg
+        |> Result.map_error (fun x -> Compiler.error_to_string x) in
+      let%ok invoked =
```
Comment (EduardoRFS) [inline src/core/contract_vm.ml:130]:
```suggestion
      (* TODO: use invoked operations for something *)
      let%ok invoked =
```

**Review by EduardoRFS — APPROVED** [review]
Please just address the comments before, but the PR looks solid to merge.

small anime girl dancing to an approved PR

**Pull request merged**